src/io_callbacks.zig

```diff
@@ -0,0 +1,902 @@
+                    try self.flush_completions(true);
+                    continue;
+                },
+                else => return err,
+            };
+            self.submitted += self.queued;
+            self.queued = 0;
+            break;
+        }
+    }
+
+    fn enqueue(self: *IO, completion: *Completion) void {
+        const sqe = self.ring.get_sqe() catch |err| switch (err) {
+            error.SubmissionQueueFull => {
+                completion.next = null;
+                self.unqueued.push(completion);
+                return;
+            },
+        };
+        self.queued += 1;
+        completion.prep(sqe);
+    }
+
+    /// This struct holds the data needed for a single io_uring operation
+    pub const Completion = struct {
+        io: *IO,
+        result: i32 = undefined,
+        next: ?*Completion = undefined,
+        operation: Operation,
+        // This is one of the usecases for c_void outside of C code and as such c_void will
+        // be replaced with anyopaque eventually: https://github.com/ziglang/zig/issues/323
+        context: ?*c_void,
+        callback: fn (context: ?*c_void, completion: *Completion, result: *const c_void) callconv(.C) void,
+
+        fn prep(completion: *Completion, sqe: *io_uring_sqe) void {
+            switch (completion.operation) {
+                .accept => |*op| {
+                    linux.io_uring_prep_accept(sqe, op.socket, &op.address, &op.address_size, op.flags);
+                },
+                .close => |op| {
+                    linux.io_uring_prep_close(sqe, op.fd);
+                },
+                .connect => |*op| {
+                    linux.io_uring_prep_connect(sqe, op.socket, &op.address, op.address_size);
+                },
+                .fsync => |op| {
+                    linux.io_uring_prep_fsync(sqe, op.fd, op.flags);
+                },
+                .openat => |op| {
+                    linux.io_uring_prep_openat(sqe, op.fd, op.path, op.flags, op.mode);
+                },
+                .read => |op| {
+                    linux.io_uring_prep_read(sqe, op.fd, op.buffer[0..buffer_limit(op.buffer.len)], op.offset);
+                },
+                .recv => |op| {
+                    linux.io_uring_prep_recv(sqe, op.socket, op.buffer, op.flags);
+                },
+                .send => |op| {
+                    linux.io_uring_prep_send(sqe, op.socket, op.buffer, op.flags);
+                },
+                .write => |op| {
+                    linux.io_uring_prep_write(sqe, op.fd, op.buffer[0..buffer_limit(op.buffer.len)], op.offset);
+                },
+            }
+            sqe.user_data = @ptrToInt(completion);
+        }
+
+        fn complete(completion: *Completion) void {
+            switch (completion.operation) {
+                .accept => {
+                    const result = if (completion.result < 0) switch (-completion.result) {
```

Comment:
This is clean. What is your feeling on this if expression? If you prefer the block that's also fine.

Reply:
Since it's just a single statement I don't think the block is necessary.